activerecord/lib/active_record/connection_adapters/abstract/transaction.rb

```diff
@@ -104,7 +104,9 @@ def initialize(connection, options, run_commit_callbacks: false)
       end
 
       def add_record(record)
-        records << record
+        unless records.map(&:object_id).include?(record.object_id)
```

Comment:
Using `map` here will allocate an array the size of `records`, which makes each additional record more expensive to add; adding N records to a transaction is now an O(N²) operation, which is a significant regression.

You could achieve the same result by telling `uniq` to compare the records by identity:

```patch
diff --git a/activerecord/lib/active_record/connection_adapters/abstract/transaction.rb b/activerecord/lib/active_record/connection_adapters/abstract/transaction.rb
index 0f2b1e85ff..fe747f8f6a 100644
--- a/activerecord/lib/active_record/connection_adapters/abstract/transaction.rb
+++ b/activerecord/lib/active_record/connection_adapters/abstract/transaction.rb
@@ -116,7 +116,7 @@ def materialized?
       end

       def rollback_records
-        ite = records.uniq
+        ite = records.uniq(&:object_id)
         while record = ite.shift
           record.rolledback!(force_restore_state: full_rollback?)
         end
@@ -127,11 +127,11 @@ def rollback_records
       end

       def before_commit_records
-        records.uniq.each(&:before_committed!) if @run_commit_callbacks
+        records.uniq(&:object_id).each(&:before_committed!) if @run_commit_callbacks
       end

       def commit_records
-        ite = records.uniq
+        ite = records.uniq(&:object_id)
         while record = ite.shift
           if @run_commit_callbacks
             record.committed!
```

Reply:
Great idea!